File: leetcode/230. Kth Smallest Element in a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        stack<TreeNode*> s;
        TreeNode* n = root;
        int i = 1;
        while (n !=NULL || !s.empty()){
            
            while (n !=  NULL) { 
                s.push(n); 
                n = n->left; 
            } 
  
            n = s.top(); 
            s.pop(); 
            if(i == k) return n->val;
            n = n->right; 
            i++;
            
        }
        return 0;
    }
    
};

File: online-training-platform/leetcode/230. Kth Smallest Element in a BST.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {

    private val list = mutableListOf<Int>()

    private fun inorder(root: TreeNode?){
        root?.let{
            inorder(it.left)
            list.add(it.`val`)
            inorder(it.right)
        }
    }

    fun kthSmallest(root: TreeNode?, k: Int): Int {
        inorder(root)
        return list[k - 1]
    }
}